### ISSUE TITLE
Unable to decode Contract Event

### ISSUE THREAD

**Issue by JeremyConnor** [polkadot-js/api#2850]
When deploying the Smart Contract, I get ```Unable to decode Contract Event``` error. 

![Screenshot from 2020-11-14 18-44-15](https://user-images.githubusercontent.com/73089748/99147756-7c335700-26a9-11eb-98bb-e69dfa7567c0.png)

The JSON that is visible here on the console also differs from the JSON that I get as a terminal output.
This is from the Browser's Console. 
```
{
  "dispatchInfo": {
    "weight": 700500000,
    "class": "Normal",
    "paysFee": "Yes"
  },
  "events": [
    {
      "phase": {
        "ApplyExtrinsic": 1
      },
      "event": {
        "index": "0x0905",
        "data": [
          "5HjpDS3KA3iAZSHNzQfyXoa72o6nadQpSjBeAT6jQrezuQf2",
          "0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d"
        ]
      },
      "topics": [
        "0x696660f4c1d1f9fc5fbc8c946c1f89e4c8b35b12ffa60d1752ef92030625f1f5"
      ]
    },
    {
      "phase": {
        "ApplyExtrinsic": 1
      },
      "event": {
        "index": "0x0000",
        "data": [
          {
            "weight": 700500000,
            "class": "Normal",
            "paysFee": "Yes"
          }
        ]
      },
      "topics": []
    }
  ],
  "status": {
    "InBlock": "0x2ffc594155a50541c9dc08008cf0b1219f051b8ebfc36c23d7215e9c15c7622e"
  },
  "contractEvents": []
}
```
Whereas the terminal output is somewhat like:
```
{
  "dispatchInfo": {
    "weight": 150000000,
    "class": "Normal",
    "paysFee": "Yes"
  },
  "events": [
    {
      "phase": {
        "ApplyExtrinsic": 1
      },
      "event": {
        "index": "0x0000",
        "data": [
          {
            "weight": 150000000,
            "class": "Normal",
            "paysFee": "Yes"
          }
        ]
      },
      "topics": []
    }
  ],
  "status": {
    "Finalized": "0x3159371db6098be8c0c9552664aa7aa2c792e6c0c7d67ebbb08aef8a0a2eb193"
  }
}
```

PS : The above two JSONs are from two different transactions, but the idea is to point out that the terminal output of the response JSON from the api-contract doesn't show full information as it is shown in the browser console.

**Comment by jacogr** [comment]
Is is possible to share the ABI and the wasm and reproduction steps (ie. Which method). Would love to replicate and fix.

**Comment by jacogr** [comment]
Actually, with the above since the event data is there, the ABI is quite possible enough to replicate in a test.

**Comment by JeremyConnor** [comment]
[AssetTransfer.zip](https://github.com/polkadot-js/api/files/5540938/AssetTransfer.zip)
Here you go. The zip file contains both the ABI and the WASM file.

**Comment by JeremyConnor** [comment]
To reproduce the JSON using the terminal, you can use the code mentioned in : 

https://github.com/polkadot-js/api/issues/2837#issuecomment-726045009

https://github.com/polkadot-js/api/issues/2837#issuecomment-727199685

Other than that, to reproduce the JSON in console, you can upload the JSON and WASM using the Polkadot UI and then after deploying the contract, call the ```register``` function. In t he register function you would need to pass an address to register that corresponding account.

**Comment by JeremyConnor** [comment]
To reproduce the JSON using the terminal, you can use the code mentioned in : 

https://github.com/polkadot-js/api/issues/2837#issuecomment-726045009

https://github.com/polkadot-js/api/issues/2837#issuecomment-727199685

Other than that, to reproduce the JSON in console, you can upload the JSON and WASM using the Polkadot UI and then after deploying the contract, call the ```register``` function. In t he register function you would need to pass an address to register that corresponding account.

**Comment by jacogr** [comment]
Ok, so you highlighted an issue on the API layer on the decoding. The linked PR fixes that. There seems to be an issue with the solang events here, I just confirmed it working on ink! with the dns sample, both API and apps UI.

So this is the difference -

ink: The format for the encoded events are `<eventId (one byte)><event data>`
solang: The format is only `<event data>`

So effectively in the sample here the data is `0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d` which is a valid publicKey (as expected), however the first byte is missing that indicates the actual event that is should be decode against. So in reality is should be `0x00<data as above>` (notice the additional first byte that indicates the "register" event type, aka enum encoding)

@seanyoung Would love your eyes on this one

**Comment by seanyoung** [comment]
I had totally missed that event decoding was already a thing, that's great! @JeremyConnor  thank you for raising this issue.

Solang is not generating correct data. I'm fixing this now.

**Comment by jacogr** [comment]
Thanks @seanyoung, appreciate it.

**Comment by seanyoung** [comment]
Should be fixed now. Solang docker images usually take a few hours to update

**Comment by JeremyConnor** [comment]
@seanyoung Will pulling the Solang Github repository work?

**Comment by jacogr** [comment]
The JS api update is in 2.7.1 that went out this morning.

**Comment by JeremyConnor** [comment]
@jacogr I tried deploying the contract using the new Solang. Now I am not getting the error ```Unable to decode Contract Event```. BUT, the previous problem of difference of result in browser console and terminal still persists. Can you suggest anything for this?

**Comment by JeremyConnor** [comment]
@jacogr I am still facing the issue of getting a different JSON result in browser's console and terminal's output.

**Comment by jacogr** [comment]
An example would be good.

**Comment by JeremyConnor** [comment]
This is what I expect, (and I am getting it in the browser console):
```
{
  "dispatchInfo": {
    "weight": 706000000,
    "class": "Normal",
    "paysFee": "Yes"
  },
  "events": [
    {
      "phase": {
        "ApplyExtrinsic": 1
      },
      "event": {
        "index": "0x0905",
        "data": [
          "5HMyuigLwpgUkbKJptuQrifB6dMkFydadcUkmUFdKoCHr4PK",
          "0x01d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d"
        ]
      },
      "topics": [
        "0x696660f4c1d1f9fc5fbc8c946c1f89e4c8b35b12ffa60d1752ef92030625f1f5"
      ]
    },
    {
      "phase": {
        "ApplyExtrinsic": 1
      },
      "event": {
        "index": "0x0000",
        "data": [
          {
            "weight": 706000000,
            "class": "Normal",
            "paysFee": "Yes"
          }
        ]
      },
      "topics": []
    }
  ],
  "status": {
    "InBlock": "0x02cb16862d06631fa08d785cb930e759d21ae72464b1cb5e25ff05e3862ceff8"
  },
  "contractEvents": [
    {
      "args": [
        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
      ],
      "event": {
        "args": [
          {
            "name": "newUser",
            "type": {
              "displayName": "AccountId",
              "info": 6,
              "type": "AccountId"
            }
          }
        ],
        "docs": [
          ""
        ],
        "identifier": "successfulRegistration",
        "index": 1
      }
    }
  ]
}
```

Whereas, this is what I am getting in the terminal output:
```
{
  "dispatchInfo": {
    "weight": 150000000,
    "class": "Normal",
    "paysFee": "Yes"
  },
  "events": [
    {
      "phase": {
        "ApplyExtrinsic": 1
      },
      "event": {
        "index": "0x0000",
        "data": [
          {
            "weight": 150000000,
            "class": "Normal",
            "paysFee": "Yes"
          }
        ]
      },
      "topics": []
    }
  ],
  "status": {
    "Finalized": "0x3159371db6098be8c0c9552664aa7aa2c792e6c0c7d67ebbb08aef8a0a2eb193"
  }
}
```

> An example of the output would be good. (And what you expect)

**Comment by jacogr** [comment]
Are you on the 2.7.1 version of the api/api-contract?

**Comment by JeremyConnor** [comment]
Yes, I updated my api and api-contract version to 2.7.1

**Comment by jacogr** [comment]
So that is the version the UI is on, so the output above is as decoded there.

So in the above example, there is no actual event that seems to come back from the node, i.e. nothing to be decoded. Only event is the normal success event, nothing else.

**Comment by JeremyConnor** [comment]
So, why am I not getting the same JSON in terminal? And how should I get it?

**Comment by jacogr** [comment]
The JSON displayed is as returned by the node. So in the second case it would seem an event is not emitted.

**Comment by JeremyConnor** [comment]
There should be a possible way to access it, right?

**Comment by jacogr** [comment]
It is not returned, so the API has nothing to decode. The contractEvents you see in the first example is the decoded `0905` event - if that is not received, the API cannot do anything.

So ensure -

- you are on the right code
- you are actually calling the right method

(Weirdly enough, the executed weight is also way off between the 2 executions, so they do not seem the same or a different path is followed)

**Comment by JeremyConnor** [comment]
I followed the documentation and use the following code. It should have worked, and I cannot find anything wrong in it.
```
const response = await contract
      .exec('register',  gasLimit, value, myAddress)
      .signAndSend(myKey,(result) => {
        if(result.status.isInBlock){          
          console.log('in a block');
        }else if (result.status.isFinalized) {
          console.log("\nResult is : ", JSON.stringify(result, null, 2));
        }
      });
```

**Comment by jacogr** [comment]
It is not the JS code, the is no event being returned from the execution. So could very well be that it follows a different path, i.e. one that does not emit in the contract.

**Comment by JeremyConnor** [comment]
> It is not the JS code, the is no event being returned from the execution. So could very well be that it follows a different path, i.e. one that does not emit in the contract. You should be able to confirm by using the UI and going to the block hash as indicated, in your case `0x3159371db6098be8c0c9552664aa7aa2c792e6c0c7d67ebbb08aef8a0a2eb193`
> 
> Looking at the weights, it seems you didn't specify enough gas possibly? (the one with events consumes almost 5 times as much as the one without)
> 
> 150000000 (without events)
> 706000000 (with events)

Thanks for pointing the difference out. I changed the gas limit, although it still didn't make any change. From your previous suggestion of trying some other method, I was thinking, is it possible to get the event information from the block hash (which I receive on successful transaction)?

**Comment by jacogr** [comment]
Sure, https://polkadot.js.org/docs/api/cookbook/blocks#how-do-i-map-extrinsics-to-their-events

(you will need to manually decode the contract events in that case, something what api-contracts does in the background, https://github.com/polkadot-js/api/blob/master/packages/api-contract/src/base/Contract.ts#L156-L164)

**Comment by JeremyConnor** [comment]
@jacogr  I tried the 1st link of the 2 that you shared. And, here is the output that I get using 
```await api.query.system.events.at(signedBlock)```(where signedBlock is the hash of the block containing the events) : 
```
[
  {
    "phase": {
      "ApplyExtrinsic": 0
    },
    "event": {
      "index": "0x0000",
      "data": [
        {
          "weight": 159133000,
          "class": "Mandatory",
          "paysFee": "Yes"
        }
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "ApplyExtrinsic": 1
    },
    "event": {
      "index": "0x0000",
      "data": [
        {
          "weight": 152000000,
          "class": "Normal",
          "paysFee": "Yes"
        }
      ]
    },
    "topics": []
  }
]

```

**Comment by jacogr** [comment]
0x0000 is System ExrinsicSuccess. 

So all it says the extrinsic made it Into the block. No other events.

**Comment by JeremyConnor** [comment]
So does that mean I cannot get the event information from this method too? If so, then I have no idea how else I am going to get one!

**Comment by jacogr** [comment]
Yes. They will be equivalent. The events are not being emitted at all. So they don’t exist.

**Comment by JeremyConnor** [comment]
As you said "Additional events are not being emitted at all." Does that mean, there is some error or Is it not yet implemented by Polkadot?

**Comment by jacogr** [comment]
It is a contracts coding issue on your side. Whichever route the code takes results in no events being emitted.

So it is purely in the contract code, not in Substrate, not in the API. Substrate doesn't get any events, it doesn't encode any events and the API cannot decode anything since well, there are no additional events. (Since events are emitted in other cases, as you linked above, we can safely assume that actual event emission as a whle works when the right codepath is followed.)

The specific transaction executed contract code follows a path which results in no events. The contract itself doesn't trap (i.e. the tx is included)

Are you sure your registration always returns events? Even when the same item is being registered? Have you tried resetting the chain (nothing in the contract) and then re-creating based on the case that did wok for you?

TL;DR This is the deployed code, none of the infrastructure.

**Comment by jacogr** [comment]
So if I look at the code in the other issue - you don’t emit anything when the user is already registered. 

This is purely the codepath being followed inside the contract itself. There are no events emitted and therefore no events on the chain.

**Comment by jacogr** [comment]
Here is your registration code as shared elsewhere -

```
function register(address _newUser) public {
        registered[_newUser] = 1;
        balance[_newUser] = 1000;

        // this method doesn't always emit - only when not already registered
        if(registered[_newUser] > 0){
            emit successfulRegistration(_newUser);
        }
    }
```

Commented above as to where I believe your issue lies.

**Comment by JeremyConnor** [comment]
> Here is your registration code as shared elsewhere -
> 
> ```
> function register(address _newUser) public {
>         registered[_newUser] = 1;
>         balance[_newUser] = 1000;
> 
>         // this method doesn't always emit - only when not already registered
>         if(registered[_newUser] > 0){
>             emit successfulRegistration(_newUser);
>         }
>     }
> ```
I made sure all of my functions were emitting events (at least when called for the first time). But, still I could not get any output on the terminal for the JSON object. I even deployed the same contract in Remix IDE and it was working perfectly fine and handling every corner case. 
When I was using the terminal, I found something peculiar. Every function was giving same output for ```await api.query.system.events.at(signedBlock)```. I made sure that each one of them was receiving different blockHash.
Here is the output of ```await api.query.system.events.at(signedBlock)```:
```
[
  {
    "phase": {
      "ApplyExtrinsic": 0
    },
    "event": {
      "index": "0x0000",
      "data": [
        {
          "weight": 159133000,
          "class": "Mandatory",
          "paysFee": "Yes"
        }
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "ApplyExtrinsic": 1
    },
    "event": {
      "index": "0x0000",
      "data": [
        {
          "weight": 152000000,
          "class": "Normal",
          "paysFee": "Yes"
        }
      ]
    },
    "topics": []
  }
]

```
This was the terminal output for every event emitted from each of the functions in the contract.

**Comment by jacogr** [comment]
Ok. There are no contract events included above. If they are not there, on-chain, nothing to decode or return.

**Comment by JeremyConnor** [comment]
But, the browser console was working perfectly fine! It was giving me the correct output. That means that the contract events were included, otherwise it would not be visible even there

**Comment by jacogr** [comment]
Yes. For those there were on-chain events. Above, there are none. It is not an API issue, it can only decode what is on-chain. There is nothing on-chain, so it does what it is supposed to - decodes what is there which doesn’t include any contract events.

**Comment by JeremyConnor** [comment]
So that means, any transaction that I make through the terminal would not be a part of 'on-chain' and hence no event?
So, my last resort to make it work would be to execute a transaction and then read the transaction to get the return values from the function using ```await contract.read(register, value, gasLimit, address)```

**Comment by jacogr** [comment]
There are no events in that block. 

The UI uses the exact same APIs as the terminal. The transaction as referenced, in that block, has no events. Nothing was emitted.

**Comment by JeremyConnor** [comment]
Thanks @jacogr  Is there any method you can suggest to get it work?

**Comment by jacogr** [comment]
I cannot debug the contract, all the API does is to decode what is received.

In this case, track why the contract is not emitting -

- do all code paths return events
- is there enough gas (for instance UI displays mega gas, so in the API itself it is 10 ^ 6 higher when passed)
- does it work when the chain is purged (apps UI and API uses exactly the same codepaths, so with the same values on both, it will be equivalent)

Sadly, nothing more I can suggest since, the API only interprets, doesn't influence and cannot interpret what is not there.

**Comment by seanyoung** [comment]
@JeremyConnor I've tried this contract, what I think is happening here is
- the `require()` statement is failing, causing a revert. This is not visible in the UI: https://github.com/paritytech/canvas-ui/issues/23 Try moving the emit to the top of the function
- events are not emitted  during rpc calls

**Comment by JeremyConnor** [comment]
@seanyoung Thanks for the insight. I could not figure this out. I will try it out and let you know. But, If I move the emit to the top of the function, then for a new user, it won't emit as the emit is being done before registering the user.

**Comment by JeremyConnor** [comment]
@seanyoung @jacogr I have changed a lot of the parts in the code. Now, the contract is emitting events in most of the cases. This zip file contains the ABI, WASM and the solidity Smart Contract.
[AssetTransfer.zip](https://github.com/polkadot-js/api/files/5555607/AssetTransfer.zip)
I also noticed many things in the UI when making transactions using the UI and when using the terminal.
In the image below, the recents events mentioned from 42 to 140 are the events emitted for the transactions made using the UI, whereas no event is seen in this section when the transaction is made using the terminal. I deployed the contract again to checkit 
![Screenshot from 2020-11-18 00-28-27](https://user-images.githubusercontent.com/73089748/99434958-75b90f80-2935-11eb-95f1-5402009be37e.png)
In the below attached image, the transactions mentioned #111 #108 #105 are made via the terminal, and visible for now. When I navigate to the other sections of the UI and come back to the Contracts page, the transactions are not visible anymore.
![Screenshot from 2020-11-18 00-28-00](https://user-images.githubusercontent.com/73089748/99435405-13144380-2936-11eb-92b8-ca150ed923c8.png)
![Screenshot from 2020-11-18 00-28-34](https://user-images.githubusercontent.com/73089748/99435408-17406100-2936-11eb-8387-1ab57650a1e0.png)

I added an asset on behalf of Alice account using the terminal, and when I moved to the UI to find the owner of the asset (I have implemented ```findOwner``` function in the smart contract), I could not see any information and not even the balance is upated. This means that none of the transaction took place! Then what is the JSON object that I am getting in the terminal as a response about?

**Comment by jacogr** [comment]
The UI uses api-contracts, it doesn't have an own implementation. So the events the UI displays is as received from api-contracts. When there are events, they are displayed. When there are none, they are not. Same between what is in "the terminal" and "on the UI". I am not the right person to help you debug your contract - what is emitted from the node is what the API displays.

If you have problems with the UI, please log it against the correct repo. (I'll save you some time, it doesn't lookup or persist "when txs took place", what is displayed there on the contracts page is all in-memory - there is no saving to DB or lookups against DB)

I'm going to close this issue specifically, the issues that were there on the integration layer were both resolved on the API level and Solang.
